packages/data-stores/src/purchases/queries/use-site-purchases.ts

```diff
@@ -0,0 +1,41 @@
+import { useQuery } from '@tanstack/react-query';
+import { useCallback } from '@wordpress/element';
+import wpcomRequest from 'wpcom-proxy-request';
+import { createPurchaseObject } from '../lib/assembler';
+import useQueryKeysFactory from './lib/use-query-keys-factory';
+import type { RawPurchase, Purchase } from '../types';
+
+interface PurchasesIndex {
+	[ purchaseId: number ]: Purchase;
+}
+
+interface Props {
+	siteId?: string | number | null;
+}
+
+function useSitePurchases( { siteId }: Props ) {
+	const queryKeys = useQueryKeysFactory();
+
+	return useQuery< RawPurchase[], Error, PurchasesIndex >( {
+		queryKey: queryKeys.sitePurchases( siteId ),
+		queryFn: async () =>
+			await wpcomRequest( {
+				path: `/sites/${ encodeURIComponent( siteId as string ) }/purchases`,
+				apiVersion: '1.3',
+			} ),
+		select: useCallback( ( data: RawPurchase[] ) => {
```

Comment:
I see, thanks for pointing me to their docs! So they're recommending wrapping a selector in `useCallback` or extracting it in case the transformation is expensive. I guess it's up to us to estimate whether that's the case.

@jsnajdr I wonder you might have any thoughts about what's considered _expensive_.

By the way I think the reducer can be optimized to not return a new object on every iteration but adding to the existing one:

```ts
data.reduce( ( acc, rawPurchase ) => {
  const purchase = createPurchaseObject( rawPurchase );
  acc[ purchase.id ] = purchase;
  return acc;
}, {} )
```

Reply:
Thanks. Let's see what a deeper investigation brings. Concerning: 

> By the way I think the reducer can be optimized to not return a new object on every iteration but adding to the existing one

I would prefer if this is handled (and recommended) at a higher level. I can imagine it being harder to control how reducers are defined inside the `select` functions. So, fewer lower-level implementation recommendations (e.g. how a `reduce` call should be implemented - which may not be a `reduce` call at all), and more higher-level ones (how a `select` function is defined) might work better. Meaning, in regards to a "best practice" guide.

Reply:
I would personally do the transformation in the `queryFn`, like `wpcomRequest().then( transform )`. Or extract the transformation to a module-level function. That's what the maintainer's blog post also recommends. The function doesn't use any local variables, so it's easy to extract and use as a constant. `useCallback` is possible, but easily avoidable.

About the `.reduce` call, the best practice nowadays is to use `Object.fromEntries`:
```js
Object.fromEntries( data.map( rawPurchase => {
  const purchase = createPurchaseObject( rawPurchase );
  return [ purchase.id, purchase ];
} ) );
```

Reply:
Thanks. It looks like the recommendations in the [post](https://tkdodo.eu/blog/react-query-data-transformations) all have their pros & cons. I'm intrigued by the `select` API / in the idea of passing finer selectors to create partial subscribers in the way the author outlines. That's pretty much what I intend to do in other work. With that, I'd rather have a lean `queryFn` that remains a lower-level fetcher. It might be the case that we use both -> to transform inside `queryFn` (I would still, however, be interested in having the actual API data inspectable in its original form) and a `select` method.

> About the .reduce call, the best practice nowadays is to use Object.fromEntries:
> 
> ```
> Object.fromEntries( data.map( rawPurchase => {
>  const purchase = createPurchaseObject( rawPurchase );
>  return [ purchase.id, purchase ];
> } ) );
> ```

Thanks for this tip. I can get used to this. Why is this a best practice btw? is it primarily performance-related? (I'm assuming it wouldn't be that different from reduce with spreading)

Reply:
> That's what the maintainer's blog post also recommends.

Hmm objectively speaking, the author outlines four recommendations with pros/cons. It looks like the last (`#3`) has the lesser caveats, but not sure if there's a preferential order there (it looks like to me, but pros/cons with each)

Reply:
Thanks, and good points/thinking all around. Really appreciate the insight. 

> Also, it's good to be aware that the `queryFn` results are what is going to be cached for the query, in IndexedDB. Therefore these data should be serializable. If you want your query to return some fancy objects, you need to use `select` to create them.

This is a good point. I'll be digging a little more into these structures.

> A very big part of the reason "why" is fashion, taste and aesthetics, so I wouldn't worry about it too much.

Totally. I think I've also used `reduce` enough that a new habit wouldn't be bad. It's also a bit more performant than reduce with spreading apparently ([this](https://www.measurethat.net/Benchmarks/Show/9268/0/objectfromentries-vs-reduce) test) - neither nearly as performant as reduce with object reuse - but that's not to my taste either)

Reply:
Thank you both for the insights. I appreciate considering different approaches and weighing concerns. It's not surprising to me that using `.reduce` with a new object on every iteration is so much slower than operating on the same object. I would've expected `Object.fromEntries` to be a bit faster though. I agree with @jsnajdr that `Object.fromEntries` is the most descriptive one and given that we don't operate on thousands of entries I wouldn't worry about it; `.reduce` is around for a very long time and hence optimized very heavily by browser engines. – I'm sorry if this is gone a little off topic @chriskmnds but I still hope it gave some valuable insights.

Reply:
So I opted to do the transformation in the `queryFn`, like @jsnajdr suggested. I think it's the cleanest approach, also for using `select` in sub-queries, as those would also work off the transformed data that way. Not sure about limitations or other caveats - they'll surface as we build out more.